rdr_service/tools/tool_libs/genomic_utils.py

```diff
@@ -1221,6 +1223,135 @@ def update_member_aw1_file_processed_id(self, mid, fid):
+
+    def process_contamination_category_using_cloud_task(self):
+        _task = None if self.gcp_env.project == 'localhost' else GCPCloudTask()
+
+        if _task is not None:
+            task_queue = 'resource-tasks'
+
+            batch_size = 100
+
+            # Get total number of batches
+            batch_total = math.ceil(len(self.member_ids) / batch_size)
+            _logger.info(f'Found {batch_total} member_id batches of size {batch_size}.')
+
+            # Setup counts
+            count = 0
+            batch_count = 0
+            batch = list()
+
+            # Add members to batch and submit cloud tasks
+            for mid in self.member_ids:
+                batch.append(mid)
+                count += 1
+
+                if count == batch_size:
+                    data = {"member_ids": batch}
+
+                    if self.args.dryrun:
+                        _logger.info("In Dryrun mode, skip submitting cloud task.")
+                    else:
+                        _task.execute('calculate_contamination_category_task',
+                                      payload=data,
+                                      queue=task_queue,
+                                      project_id=self.gcp_env.project)
+
+                    batch_count += 1
+                    _logger.info(f'Task created for batch {batch_count}')
+
+                    # Reset counts
+                    batch.clear()
+                    count = 0
+
+            # Submit remainder in last batch
+            if count > 0:
+                batch_count += 1
+                data = {"member_ids": batch}
+                _task.execute('calculate_contamination_category_task',
+                              payload=data,
+                              queue=task_queue,
+                              project_id=self.gcp_env.project)
+
+            _logger.info(f'Submitted {batch_count} tasks.')
+
+        return 0
+
+    def process_contamination_category_locally(self):
+
+        genomic_ingester = GenomicFileIngester(job_id=GenomicJob.RECALCULATE_CONTAMINATION_CATEGORY)
+
+        for mid in self.member_ids:
+
+            # Get genomic_set_member and gc metric objects
+            with self.dao.session() as s:
+                record = s.query(GenomicSetMember, GenomicGCValidationMetrics).filter(
+                    GenomicSetMember.id == mid,
+                    GenomicSetMember.collectionTubeId != None,
+                    GenomicGCValidationMetrics.genomicSetMemberId == mid
+                ).one_or_none()
+
+                if record is not None:
+                    # calculate new contamination category
+                    contamination_category = genomic_ingester.calculate_contamination_category(
+                        record.GenomicSetMember.collectionTubeId,
+                        float(record.GenomicGCValidationMetrics.contamination),
+                        record.GenomicSetMember
+                        )
```

Comment:
I don't know if this might make a difference for your workflow, but the `calculate_contamination_category` method modifies the database by adding samples to the contaminated table if they should be there, regardless of whether the tool is being run as a dry-run or not. Maybe it's fine as it is, but I wanted to make sure you were aware of it.

Reply:
Good observation, I've updated the logic to only call `calculate_contamination_category` when `dryrun` is false.